File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if(height.size() < 2)
            return 0;
        int l = 0, r = height.size() - 1, lmax = height[0], rmax  = height[height.size()-1], area = 0;
        while(l < r){
            if(height[l] < height[r]){
                lmax = max(height[l], lmax);
                area += lmax - height[l];
                l++;
            }else{
                rmax = max(height[r], rmax);
                area += rmax - height[r];
                r--;
            }
            // cout<<lmax<<" "<<rmax<<endl;
            // cout<<area<<endl;
        }
        return area;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        // [4,2,0,3,2,5]
        //  0,4,4,4,4,4
        //  5,5,5,5,5,0
      //    0,2,4,1,2,0
        // min(leftMax, rightMax) - height[i];
        int[] leftMax = new int[n];
        int[] rightMax = new int[n];

        for(int i = 1; i < n; i++){
            leftMax[i] = Math.max(height[i-1], leftMax[i-1]);
            rightMax[n-i-1] = Math.max(height[n-i], rightMax[n-i]);
        }
        int ans = 0;
        for(int i = 0; i < n; i++){
            int waterCapacity = Math.min(leftMax[i], rightMax[i])*1;
            if(waterCapacity > height[i])
                ans += waterCapacity - height[i];
        }
        return ans;
        
    }
}